File: golang-backend/internal/handlers/rolehandler/dto.go
package rolehandler

import "github.com/aportela/doneo/internal/handlers"

type permissionsFlags struct {
	AllowUpdateProject bool `json:"allowUpdateProject"`
	AllowDeleteProject bool `json:"allowDeleteProject"`
	AllowViewProject   bool `json:"allowViewProject"`
	AllowAddTask       bool `json:"allowAddTask"`
	AllowUpdateTask    bool `json:"allowUpdateTask"`
	AllowDeleteTask    bool `json:"allowDeleteTask"`
	AllowViewTask      bool `json:"allowViewTask"`
}

type addRequest struct {
	Name        string           `json:"name"`
	Permissions permissionsFlags `json:"permissions"`
}

type updateRequest struct {
	ID          string           `json:"id"`
	Name        string           `json:"name"`
	Permissions permissionsFlags `json:"permissions"`
}

type filterRequest struct {
	Name *string `json:"name"`
}

type searchRequest struct {
	Pager  handlers.PagerRequest `json:"pager"`
	Order  handlers.OrderRequest `json:"order"`
	Filter *filterRequest        `json:"filter"`
}

type RoleBaseResponse struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type RoleResponse struct {
	RoleBaseResponse
	Permissions permissionsFlags `json:"permissions"`
}

type searchBaseResponse struct {
	Roles []RoleBaseResponse `json:"roles"`
}

type searchResponse struct {
	Roles []RoleResponse         `json:"roles"`
	Pager handlers.PagerResponse `json:"pager"`
}

File: vite-vue-ts-frontend/src/modules/roles/types/dto.ts
import type {
  PagerRequest,
  PagerResponse,
} from "../../../shared/types/dto/pager";
import type { Order } from "../../../shared/types/dto/order";

export type Permissions = {
  allowCreate: boolean;
  allowUpdate: boolean;
  allowDelete: boolean;
  allowView: boolean;
  allowList: boolean;
  allowExecute: boolean;
};

export type AddRequest = {
  name: string;
  permissions: Permissions;
};

export type UpdateRequest = {
  id: string;
  name: string;
  permissions: Permissions;
};

type SearchRequestFilter = {
  name?: string;
  permissions?: {
    allowCreate?: boolean;
    allowUpdate?: boolean;
    allowDelete?: boolean;
    allowView?: boolean;
    allowList?: boolean;
    allowExecute?: boolean;
  };
};

export type SearchRequest = {
  pager: PagerRequest;
  order: Order;
  filter?: SearchRequestFilter;
};

export type RoleResponse = {
  id: string;
  name: string;
  permissions: Permissions;
};

export type SearchResponse = {
  roles: RoleResponse[];
  pager: PagerResponse;
};
